app/src/main/java/net/squanchy/search/view/ItemsAdapter.java

```diff
@@ -1,41 +1,135 @@
+        } else {
+            return ViewTypeId.SPEAKER;
+        }
+    }
+
+    long itemIdAtAbsolutePosition(int position) {
+        ensurePositionExists(position);
+
+        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResults.events());
+        if (totalEventItemsCount > 0 && position < totalEventItemsCount) {
+            if (position == 0) {
+                return ITEM_ID_EVENTS_HEADER;
+            } else {
+                return searchResults.events().get(position - 1).numericId();
+            }
+        } else {
+            int adjustedPosition = position - totalEventItemsCount;
+            // We checked position is in range so it MUST be a speaker
+            if (adjustedPosition == 0) {
+                return ITEM_ID_SPEAKERS_HEADER;
+            } else {
+                return searchResults.speakers().get(adjustedPosition - 1).numericId();
+            }
         }
-        return ViewTypeId.SPEAKER;
+    }
+
+    Speaker speakerAtAbsolutePosition(int position) {
+        ensurePositionExists(position);
+
+        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResults.events());
+        if (totalEventItemsCount > 0 && position < totalEventItemsCount) {
+            throw new IndexOutOfBoundsException("No speaker at position " + position + ", that is supposed to be in the events sublist");
+        }
+
+        int adjustedPosition = position - totalEventItemsCount;
+        // We checked position is in range so it MUST be a speaker
+        if (adjustedPosition > 0) {
+            return searchResults.speakers().get(adjustedPosition - 1);
+        } else {
+            throw new IndexOutOfBoundsException("No speaker at position " + position + ", that is supposed to be the speakers header");
+        }
+    }
+
+    HeaderType headerTypeAtAbsolutePosition(int position) {
+        ensurePositionExists(position);
+
+        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResults.events());
+        if (totalEventItemsCount > 0) {
+            if (position == 0) {
+                return HeaderType.EVENTS;
+            } else if (position < totalEventItemsCount) {
+                throw new IndexOutOfBoundsException("No header at position " + position + ", that is supposed to be an event");
+            }
+        }
+
+        int adjustedPosition = position - totalEventItemsCount;
+        // We checked position is in range so it MUST be a speaker
+        if (adjustedPosition == 0) {
+            return HeaderType.SPEAKERS;
+        } else {
+            throw new IndexOutOfBoundsException("No header at position " + position + ", that is supposed to be a speaker");
+        }
+    }
+
+    private void ensurePositionExists(int position) {
+        if (position < 0 || position >= totalItemsCount()) {
+            String message = String.format(Locale.ROOT, "Position %d is not valid, must be [0, totalItemsCount)", position);
```

Comment:
can you substitute `totalItemsCount` in the message with the actual items count?